lib/fluent/plugin/out_elasticsearch.rb

```diff
@@ -354,7 +370,36 @@ def send_bulk(data)
     begin
       response = client.bulk body: data
       if response['errors']
-        log.error "Could not push log to Elasticsearch: #{response}"
+        errors = Hash.new(0)
+        errors_bad = 0
+        errors_unrecognized = 0
+        # Count up the individual error types returned for each item
+        response['items'].each do |item|
+          begin
+            if [429, 500].include?(item['create']['status'])
+              errors[item['create']['error']['type']] += 1
+            else
+              errors_unrecognized += 1
+            end
+          rescue Exception
+            errors_bad += 1
+          end
+        end
+        if errors_bad > 0
+          raise ElasticsearchVersionMismatch, "Unable to parse error response from Elasticsearch, likely an API version mismatch  #{response}"
+        end
+        if errors_unrecognized > 0
+          raise UnrecognizedElasticsearchError, "Unrecognized elasticsearch errors returned, retrying  #{response}"
+        end
+        errors.each_key do |key|
+          if 'out_of_memory_error' == key
+            raise ElasticsearchOutOfMemory, "Elasticsearch has exhausted its heap, retrying"
```

Comment:
Unfortunately, there are two retry loops, the one below for `host_unreachable_exceptions` and the retry loop for errors in the output thread in the main fluentd code itself.